File: src/main/java/hiiragi283/material/api/registry/HiiragiRegistry.java
package hiiragi283.material.api.registry;

import org.apache.logging.log4j.LogManager;
import org.apache.logging.log4j.Logger;
import org.jetbrains.annotations.NotNull;

import javax.annotation.ParametersAreNonnullByDefault;
import java.util.Collection;
import java.util.LinkedHashMap;
import java.util.Optional;

@ParametersAreNonnullByDefault
public class HiiragiRegistry<K, V> {

    private boolean isUnmodifiable = false;
    private final LinkedHashMap<K, V> REGISTRY = new LinkedHashMap<>();
    private final String name;
    protected final Logger LOGGER;

    public HiiragiRegistry(String name) {
        this.name = name;
        this.LOGGER = LogManager.getLogger(name);
    }

    @NotNull
    public String getName() {
        return name;
    }

    @NotNull
    public Collection<V> getValues() {
        return REGISTRY.values();
    }

    @NotNull
    public Optional<V> getValue(K key) {
        return Optional.ofNullable(REGISTRY.get(key));
    }

    public void register(K key, V value) {
        if (isUnmodifiable) {
            LOGGER.error("Cannot register any values with this registry!");
        } else if (REGISTRY.containsKey(key)) {
            LOGGER.error("The key: " + key + " is already registered!");
        } else {
            REGISTRY.putIfAbsent(key, value);
        }
    }

    public void setUnmodifiable() {
        isUnmodifiable = true;
    }

}

File: src/main/kotlin/hiiragi283/material/api/registry/HiiragiRegistry.kt
package hiiragi283.material.api.registry

open class HiiragiRegistry<K, V>(val name: String, private val mutable: Boolean = false) {

    //    Lock    //

    protected var isLocked: Boolean = false

    fun lock() {
        isLocked = true
    }

    //    Registration    //

    protected val registry: LinkedHashMap<K, V> = linkedMapOf()

    fun getEntries(): List<Pair<K, V>> = registry.toList()

    fun getValues(): Collection<V> = registry.values

    operator fun get(key: K): V? = registry[key]

    protected val reversed: LinkedHashMap<V, K> = linkedMapOf()

    fun getReversedEntries(): List<Pair<V, K>> = reversed.toList()

    @Synchronized
    internal operator fun <T : V> set(key: K, value: T): T = when {
        isLocked -> throw IllegalStateException("[$name] This registry is locked!")
        registry.containsKey(key) && !mutable -> throw IllegalStateException("[$name] The key: $key is already registered!")
        else -> {
            registry[key] = value
            value
        }
    }

    @Synchronized
    fun remove(key: K): V? = when {
        isLocked -> throw IllegalStateException("[$name] This registry is locked!")
        !mutable -> throw IllegalStateException("[$name] This registry cannot remove entry!")
        !registry.containsKey(key) -> throw IllegalStateException("[$name] The key: $key is not registered!")
        else -> registry.remove(key)
    }

    //    Operation    //

    operator fun contains(key: K): Boolean = registry.containsKey(key)

}